File: accbot-android/app/src/main/java/com/accbot/dca/domain/usecase/ExportTransactionsToCsvUseCase.kt
package com.accbot.dca.domain.usecase

import com.accbot.dca.data.local.TransactionDao
import com.accbot.dca.data.local.TransactionEntity
import com.accbot.dca.presentation.utils.DateFormatters
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import javax.inject.Inject

/**
 * Result of CSV export operation.
 */
sealed class CsvExportResult {
    /**
     * Export successful with CSV content and suggested filename.
     */
    data class Success(
        val csvContent: String,
        val suggestedFileName: String
    ) : CsvExportResult()

    /**
     * Export failed with error message.
     */
    data class Error(val message: String) : CsvExportResult()
}

/**
 * Use case for exporting transactions to CSV format.
 * Extracts CSV generation logic from HistoryViewModel for better testability
 * and to remove Context dependency from ViewModel.
 *
 * The actual file writing and sharing is handled by the UI layer,
 * which has access to Context.
 */
class ExportTransactionsToCsvUseCase @Inject constructor(
    private val transactionDao: TransactionDao
) {
    companion object {
        private val fileNameFormatter = DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss")
            .withZone(ZoneId.systemDefault())
    }

    /**
     * Generate CSV content from all transactions.
     *
     * @return CsvExportResult with content and filename, or error
     */
    suspend fun execute(): CsvExportResult {
        return try {
            val transactions = transactionDao.getAllTransactionsOnce()

            if (transactions.isEmpty()) {
                return CsvExportResult.Error("No transactions to export")
            }

            val csvContent = generateCsv(transactions)
            val fileName = "accbot_transactions_${fileNameFormatter.format(Instant.now())}.csv"

            CsvExportResult.Success(
                csvContent = csvContent,
                suggestedFileName = fileName
            )
        } catch (e: Exception) {
            CsvExportResult.Error(e.message ?: "Failed to export transactions")
        }
    }

    /**
     * Escape a CSV field value, quoting it if it contains special characters.
     */
    private fun escapeCsvField(value: String): String {
        return if (value.contains(',') || value.contains('"') || value.contains('\n') || value.contains('\r')) {
            "\"${value.replace("\"", "\"\"")}\""
        } else {
            value
        }
    }

    /**
     * Generate CSV string from transactions.
     */
    private fun generateCsv(transactions: List<TransactionEntity>): String {
        return buildString {
            // CSV header
            appendLine("Date,Exchange,Crypto,Fiat,Crypto Amount,Fiat Amount,Price,Fee,Fee Asset,Status,Order ID,Error")

            // Data rows
            transactions.forEach { tx ->
                append(DateFormatters.isoDateTime.format(tx.executedAt))
                append(",")
                append(escapeCsvField(tx.exchange.displayName))
                append(",")
                append(escapeCsvField(tx.crypto))
                append(",")
                append(escapeCsvField(tx.fiat))
                append(",")
                append(tx.cryptoAmount.toPlainString())
                append(",")
                append(tx.fiatAmount.toPlainString())
                append(",")
                append(tx.price.toPlainString())
                append(",")
                append(tx.fee.toPlainString())
                append(",")
                append(escapeCsvField(tx.feeAsset))
                append(",")
                append(escapeCsvField(tx.status.name))
                append(",")
                append(escapeCsvField(tx.exchangeOrderId ?: ""))
                append(",")
                append(escapeCsvField(tx.errorMessage ?: ""))
                appendLine()
            }
        }
    }
}

File: accbot-ios/AccBot/Domain/UseCase/ExportTransactionsToCsvUseCase.swift
import Foundation

enum CsvExportResult {
    case success(csvContent: String, suggestedFileName: String)
    case error(String)
}

/// Exports all transactions to CSV format with proper escaping.
final class ExportTransactionsToCsvUseCase {
    private let transactionDao: TransactionDao

    init(transactionDao: TransactionDao) {
        self.transactionDao = transactionDao
    }

    func execute() throws -> CsvExportResult {
        let transactions = try transactionDao.getAllTransactionsOnce()

        guard !transactions.isEmpty else {
            return .error("No transactions to export")
        }

        let csvContent = generateCsv(transactions)
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd_HHmmss"
        let fileName = "accbot_transactions_\(formatter.string(from: Date())).csv"

        return .success(csvContent: csvContent, suggestedFileName: fileName)
    }

    private func generateCsv(_ transactions: [Transaction]) -> String {
        var csv = "Date,Exchange,Crypto,Fiat,Crypto Amount,Fiat Amount,Price,Fee,Fee Asset,Status,Order ID,Error,Warning\n"

        let dateFormatter = ISO8601DateFormatter()
        dateFormatter.formatOptions = [.withInternetDateTime]

        for tx in transactions {
            csv += dateFormatter.string(from: tx.executedAt)
            csv += ","
            csv += escapeCsvField(tx.exchange.displayName)
            csv += ","
            csv += escapeCsvField(tx.crypto)
            csv += ","
            csv += escapeCsvField(tx.fiat)
            csv += ","
            csv += "\(tx.cryptoAmount)"
            csv += ","
            csv += "\(tx.fiatAmount)"
            csv += ","
            csv += "\(tx.price)"
            csv += ","
            csv += "\(tx.fee)"
            csv += ","
            csv += escapeCsvField(tx.feeAsset)
            csv += ","
            csv += escapeCsvField(tx.status.rawValue)
            csv += ","
            csv += escapeCsvField(tx.exchangeOrderId ?? "")
            csv += ","
            csv += escapeCsvField(tx.errorMessage ?? "")
            csv += ","
            csv += escapeCsvField(tx.warningMessage ?? "")
            csv += "\n"
        }

        return csv
    }

    private func escapeCsvField(_ value: String) -> String {
        if value.contains(",") || value.contains("\"") || value.contains("\n") || value.contains("\r") {
            return "\"\(value.replacingOccurrences(of: "\"", with: "\"\""))\""
        }
        return value
    }
}
